homeassistant/components/smarttub/sensor.py

```diff
@@ -36,17 +44,69 @@ async def async_setup_entry(hass, entry, async_add_entities):
 class SmartTubSensor(SmartTubEntity):
     """Generic and base class for SmartTub sensors."""
 
-    def __init__(self, coordinator, spa, sensor_name, spa_status_key):
+    def __init__(self, coordinator, spa, sensor_name, attr_name):
         """Initialize the entity."""
         super().__init__(coordinator, spa, sensor_name)
-        self._spa_status_key = spa_status_key
+        self._attr_name = attr_name
 
     @property
     def _state(self):
         """Retrieve the underlying state from the spa."""
-        return self.get_spa_status(self._spa_status_key)
+        return getattr(self.spa_status, self._attr_name)
 
     @property
     def state(self) -> str:
         """Return the current state of the sensor."""
+        if isinstance(self._state, Enum):
+            return self._state.name.lower()
         return self._state.lower()
+
+
+class SmartTubPrimaryFiltrationCycle(SmartTubSensor):
+    """The primary filtration cycle."""
+
+    def __init__(self, coordinator, spa):
+        """Initialize the entity."""
+        super().__init__(
+            coordinator, spa, "primary filtration cycle", "primary_filtration"
+        )
+
+    @property
+    def state(self) -> str:
+        """Return the current state of the sensor."""
+        return self._state.status.name.lower()
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes."""
+        state = self._state
+        return {
+            ATTR_DURATION: state.duration,
+            ATTR_LAST_UPDATED: state.last_updated.isoformat(),
+            ATTR_MODE: state.mode.name.lower(),
+            ATTR_START_HOUR: state.start_hour,
```

Comment:
What does the start hour return for value? Is it a timestamp?

Reply:
It is an integer from 0-23 representing the time of day when the cycle will run. Only the hour can be configured, the minute is always 0. I'll add a comment.

Reply:
Addressed in https://github.com/home-assistant/core/pull/47192

Reply:
One thing to consider is to create two separate sensor entities with device class timestamp representing the start and end times of the current cycle.

https://developers.home-assistant.io/docs/core/entity/sensor#available-device-classes

This may make it easier to automate around these measurements.

Reply:
Thanks for your feedback.

When the cycle begins or ends, the state is updated (from "inactive" to "active" to "inactive" again), so if a user wants to trigger automations relative to the cycle, that's how I would expect them to do it. Do you think it would be useful to expose the expected start/end time of the cycle before it has begun? Why would it be preferable to have this in two separate entities rather than state attributes? I'm a bit unclear on when to use one vs. the other.

Note that I want to add services (and/or device actions?) to allow the user to change the settings for each cycle: https://github.com/home-assistant/core/pull/46980

Reply:
Yeah, having a timestamp sensor with the next cycle time gives more options that acting just when the state changes.

Generally we prefer separate sensors compared to state attributes since it's easier to use that state directly in an automation and it has more information eg unit of measurement when available.

Reply:
OK, that makes sense. I'll add that to my todo list.